src/operator/nn/mkldnn/mkldnn_base-inl.h

```diff
@@ -147,6 +147,23 @@ static inline bool MKLDNNEnvSet() {
   return is_mkldnn_enabled;
 }
 
+static inline int GetMKLDNNCacheSize() {
+  static int mkldnn_cache_size = dmlc::GetEnv("MXNET_MKLDNN_CACHE_SIZE", -1);
+  return mkldnn_cache_size;
+}
+
+// TODO(alex): (MXNET-1075) Will remove env variable and calculate cache size during runtime
+template<typename S, typename I, typename H>
+static typename std::unordered_map<S, I, H>::iterator AddToCache(
+    std::unordered_map<S, I, H>* cache, const S &key, const I &item) {
+  int mkldnn_cache_size = GetMKLDNNCacheSize();
+  if (mkldnn_cache_size != -1 && static_cast<int>(cache.size()) > mkldnn_cache_size)
+    cache->erase(cache.begin());
```

Comment:
This is a little crude to me. Can we have a LRU algorithm here? https://en.wikipedia.org/wiki/Cache_replacement_policies#LRU

Reply:
There're 2 approach reasonable to me.
1. Permanently disable cache and remove the overhead of key hashing. By this way, we can get best performance and minimal memory usage for the cache always miss scenarios.
2. Maintain cache size as user expected, and kick off an old one when reaching the memory limitation. Then we can expect cache may hit in future. Just like what cpu cache behaves. 

I don't know why you said 

> if the future inputs are consistently not in the cache then we need to constant remove one element every time which is expensive.

Removing an item from hash map has constant time complexity, which is much less than building key hashing. If you think it's expensive, then we should move to first approach(not hashing key at all).